server/remote_cache/byte_stream_server/byte_stream_server.go

```diff
@@ -202,10 +202,27 @@ func (s *ByteStreamServer) Read(req *bspb.ReadRequest, stream bspb.ByteStream_Re
 // service was able to commit and whether the service views the resource as
 // `complete` or not.
 
-type writeState struct {
+// WriteState enapsulates an on-going ByteStream write to a cache, freeing
+// the caller of having to manage writing and committing-to the cache, tracking
+// cache hits, verifying checksums, etc. Here is how it must be used:
+//   - A new WriteState may be obtained by providing the first frame of the
+//     stream to the BeginWrite() function. This function will return a new
+//     WriteState, or an error.
+//   - If a WriteState is returned from BeginWrite(), the function returned from
+//     WriteState.Cleanup() must be deferred to free system resources.
+//   - Each subsequent frame should be passed to WriteState.Write(), which will
+//     return an error on error (note: io.EOF indicates the cache believes the
+//     write is finished), or an optional WriteResponse that should be sent to
+//     the client if the client indicated the write is finished. This function
+//     will return (nil, nil) if the frame was processed successfully, but the
+//     write is not finished yet.
+type WriteState struct {
```

Comment:
since we're renaming this I think it'd be better to call this byte_stream_server.WriteHandler.

Reply:
Good idea, done